File: challenges/linked_list/go/insertion_sortlist.go
package main

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func insertionSortList(head *ListNode) *ListNode {
	if head == nil {
		return head
	}

	dummy := &ListNode{}
	curr := head
	var next *ListNode
	var prev *ListNode = dummy
	for curr != nil {
		next = curr.Next
		if prev.Val > curr.Val {
			prev = dummy
		}

		for prev.Next != nil && prev.Next.Val < curr.Val {
			prev = prev.Next
		}

		curr.Next, prev.Next, curr = prev.Next, curr, next

	}

	return dummy.Next
}

func main() {
	lst := &ListNode{Val: 4, Next: &ListNode{Val: 2, Next: &ListNode{Val: 1, Next: &ListNode{Val: 3}}}}
	insertionSortList(lst)
	// print
	for lst != nil {
		println(lst.Val)
		lst = lst.Next
	}
}

File: challenges/arrays/dart/insertion_sortlist.dart
/**
 * Definition for singly-linked list.
 * class ListNode {
 *   int val;
 *   ListNode? next;
 *   ListNode([this.val = 0, this.next]);
 * }
 */
class Solution {
    // Approach 1: Sort array and set the values in linked list O(nlogn)    
  ListNode? insertionSortList(ListNode? head) {
        List<int> nums = [];
            ListNode? temp = head;
            while (temp != null) {
                nums.add(temp.val);
                temp = temp.next;
            }
            nums.sort();
            temp = head;
            int i = 0;
            while (temp != null) {
                temp.val = nums[i++];
                temp = temp.next;
            }
            return head;
  }
   // Approach 2: Insertion Sort O(n^2)
    ListNode? insertionSortList(ListNode? head) {
 if (head == null) {
      return head;
    }
      ListNode? dummy = ListNode(0);
        ListNode? curr = head;
        ListNode? next = null;
        while(curr !=null ){
             next = curr.next;
             ListNode? prev = dummy;
                while(prev!.next != null && prev.next!.val < curr.val){
                    prev = prev.next;
                }
                curr.next = prev.next;
                prev.next = curr;
                curr = next;
            
        }
        return dummy.next;

}